tests/models/test_dag.py

```diff
@@ -979,34 +979,18 @@ def test_duplicate_task_ids_not_allowed_with_dag_context_manager(self):
 
         self.assertEqual(dag.task_dict, {op1.task_id: op1})
 
-        # Also verify that DAGs with duplicate task_ids don't raise errors
-        with DAG("test_dag_1", start_date=DEFAULT_DATE) as dag1:
-            op3 = DummyOperator(task_id="t3")
-            op4 = BashOperator(task_id="t4", bash_command="sleep 1")
-            op3 >> op4
-
-        self.assertEqual(dag1.task_dict, {op3.task_id: op3, op4.task_id: op4})
-
     def test_duplicate_task_ids_not_allowed_without_dag_context_manager(self):
         """Verify tasks with Duplicate task_id raises error"""
         with self.assertRaisesRegex(
             DuplicateTaskIdFound, "Task id 't1' has already been added to the DAG"
         ):
-            dag = DAG("test_dag", start_date=DEFAULT_DATE)
+            dag = DAG("test_dag", default_args={'start_date': DEFAULT_DATE})
```

Comment:
Moved the `start_date` to `default_args` as without it we were getting lucky where op1.start_date was set but op2.start_date was None  :

```python
In [3]: from airflow.operators.dummy_operator import DummyOperator; from airflow.models import DAG

In [4]: from airflow.utils import timezone

In [5]: dag = DAG("test_dag", default_args=dict(start_date=timezone.datetime(2019, 1, 1)))

In [6]: with dag:
   ...:     op1 = DummyOperator(task_id="t1")
   ...:     op2 = DummyOperator(task_id="t1")
   ...:

In [7]: op1.start_date
Out[7]: datetime.datetime(2019, 1, 1, 0, 0, tzinfo=<TimezoneInfo [UTC, GMT, +00:00:00, STD]>)

In [8]: op2.start_date
Out[8]: datetime.datetime(2019, 1, 1, 0, 0, tzinfo=<TimezoneInfo [UTC, GMT, +00:00:00, STD]>)

In [9]: dag = DAG("test_dag", start_date=timezone.datetime(2019, 1, 1))

In [10]: with dag:
    ...:     op1 = DummyOperator(task_id="t1")
    ...:     op2 = DummyOperator(task_id="t1")
    ...:

In [11]: op1.start_date
Out[11]: datetime.datetime(2019, 1, 1, 0, 0, tzinfo=<Timezone [UTC]>)

In [12]: op2.start_date
<None>
```

Reply:
I don't get why this change is needed -- we don't look at _any_ properties anymore, just is it the same object

Reply:
This is just to stop regression, without this change if we go back to `!=` check all the tests would still pass. This change will fail the test

Reply:
The `default_args` change shouldn't be needed to prevent regressions.

Reply:
It _should not_, true but it does. If you check the iPython bit above, when the `default_args` are passed to the tasks they have `start_date` hence the `op1==op2` but when `start_date` is directly passed as `dag` argument `op1 != op2` because `op1.start_date` get's the start_date of the DAG while the `op2.start_date` is `None`.

So if we don't have a test with `start_date` set in `default_args`, if someone changes `op1 is not op2` check to `op1 != op2` check, all the test would still PASS.

Reply:
Surely `DummyOp('t1')` and `DummyOp('t2')` would fail the this test -- cos it would then not raise an error.

Reply:
It won't because of the second block in Python. See the following example:

```
In [9]: dag = DAG("test_dag", start_date=timezone.datetime(2019, 1, 1))

In [10]: with dag:
    ...:     op1 = DummyOperator(task_id="t1")
    ...:     op2 = DummyOperator(task_id="t1")
    ...:

In [11]: op1.start_date
Out[11]: datetime.datetime(2019, 1, 1, 0, 0, tzinfo=<Timezone [UTC]>)

In [12]: op2.start_date
<None>
```

i.e `op1 != op2` so it would still raise the error as `start_date` is not the same for both and `start_date` is part of `BaseOperator._comps`:

https://github.com/apache/airflow/blob/f1dc2e0b0e358582c1df0cc07a5cc95fa721dc44/airflow/models/baseoperator.py#L270-L290

https://github.com/apache/airflow/blob/f1dc2e0b0e358582c1df0cc07a5cc95fa721dc44/airflow/models/baseoperator.py#L450-L453

Reply:
Updating the place where I raise the error so that `start_date` in `default_args` doesn't make any difference